### PR TITLE
Feature/370 replace template engine icons 6.3.x

### PR DESCRIPTION
**Pull request by pixeljunkie**
| Q             | A
| ------------- | ---
| Branch        | 6.3.x
| Bug fix?      | no
| New feature?  | yes
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | yes
| Fixed issues  | chameleon-system/chameleon-system#370
| License       | MIT

### CODE REVIEW

File: src/CoreBundle/private/modules/CMSModuleChooser/views/readonly.view.php [hunk 1]
```diff
@@ -1,29 +1,41 @@
...
 <div style="margin-top: 4px;position:relative;z-index:1000">
     <div style="border-color:#<?=$oModuleInstanceColorState; ?>" class="CMSModuleChooserReadOnly"
          spotname="<?=$menuPrefix; ?>">
-        <?php if (!is_null($data['oModule'])) {
-    ?>
+        <?php
+        if (null !== $oModule) {
+            $iconFontCssClass = $oModule->fieldIconFontCssClass;
+            if ('' === $iconFontCssClass) {
+                $iconFontCssClass = '';
```
Comment (umulmrum) [inline src/CoreBundle/private/modules/CMSModuleChooser/views/readonly.view.php]: I think this is a missing fallback.

File: src/CoreBundle/private/modules/CMSModuleChooser/views/standard.view.php [hunk 2]
```diff
@@ -12,67 +18,72 @@
...
                onclick="return false;"
-               style="display: block; font-weight: bold; color: #fff; background: #20a8d8 url(<?=TGlobal::GetStaticURLToWebLib('/images/icons/layout_edit.png'); ?>) no-repeat scroll 5px center"><span><?=$translator->trans('chameleon_system_core.template_engine.spot_menu_headline', array(), TranslationConstants::DOMAIN_BACKEND); ?></span></a>
+               ><i class="fas fa-edit"></i> <?=$translator->trans('chameleon_system_core.template_engine.spot_menu_headline', array(), TranslationConstants::DOMAIN_BACKEND); ?></a>
         </div>
-        <?php if (!is_null($data['oModule'])) {
+        <?php
+        if (null !== $oModule) {
+            $iconFontCssClass = $oModule->fieldIconFontCssClass;
+            if ('' === $iconFontCssClass) {
+                $iconFontCssClass = '';
```
Comment (umulmrum) [inline src/CoreBundle/private/modules/CMSModuleChooser/views/standard.view.php]: Missing fallback.

File: src/CoreBundle/Resources/views/snippets/CMSModuleChooser/headerjs.html.twig [hunk 3]
```diff
@@ -2,6 +2,7 @@
 
 <link href="{{ themePath }}/css/cms_user_style/main.css" rel="stylesheet" type="text/css" />
 <link href="{{ resourcePath }}/css/modules/templateengineeditor.css" rel="stylesheet" type="text/css" />
+<link href="{{ resourcePath }}/iconFonts/fontawesome-free-5.8.1/css/all.css" media="screen" rel="stylesheet" type="text/css" />
```
Comment (umulmrum) [inline src/CoreBundle/Resources/views/snippets/CMSModuleChooser/headerjs.html.twig]: In Chameleon 6.3.0 the correct version is 5.5.0.

File: src/CoreBundle/Bridge/Chameleon/Migration/Script/update-1554106531.inc.php [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+  ->setWhereEquals([
+      'id' => '6a8eb0c7-bb06-e930-f29b-d6e9073fc1d2',
+  ])
+;
+TCMSLogChange::update(__LINE__, $data);
+
+$data = TCMSLogChange::createMigrationQueryData('cms_field_conf', 'en')
+    ->setFields([
+//        'name' => 'icon_font_css_class',
+        'translation' => 'Icon-Font CSS-Class',
```
Comment (umulmrum) [inline src/CoreBundle/Bridge/Chameleon/Migration/Script/update-1554106531.inc.php]:
```suggestion
        'translation' => 'Icon font CSS class',
```

File: src/CoreBundle/Bridge/Chameleon/Migration/Mapping/IconMapping.php [hunk 5]
```diff
@@ -0,0 +1,192 @@
...
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace ChameleonSystem\CoreBundle\Bridge\Chameleon\Migration\Mapping;
+
+class IconMapping
+{
+    private $iconMapping = [
```
[the remark this answers is not in the document]
Comment (pixeljunkie) [reply src/CoreBundle/Bridge/Chameleon/Migration/Mapping/IconMapping.php]: done

File: src/CoreBundle/Bridge/Chameleon/Migration/Script/update-1554106531.inc.php [hunk 6]
```diff
@@ -0,0 +1,83 @@
+<h1>Build #1554106531</h1>
+<h2>Date: 2019-04-01</h2>
+<div class="changelog">
+    - add CSS icon font field, deprecate old icon field in frontend modules
+</div>
+<?php
+
+$fieldId = TCMSLogChange::GetTableFieldId(TCMSLogChange::GetTableId('cms_tpl_module'), 'icon_list');
+
+$data = TCMSLogChange::createMigrationQueryData('cms_field_conf', 'de')
```
[the remark this answers is not in the document]
Comment (pixeljunkie) [reply src/CoreBundle/Bridge/Chameleon/Migration/Script/update-1554106531.inc.php]: done

File: src/CoreBundle/Bridge/Chameleon/Migration/Script/update-1554106531.inc.php [hunk 7]
```diff
@@ -0,0 +1,83 @@
...
+  ])
+;
+TCMSLogChange::update(__LINE__, $data);
+
+$query = "ALTER TABLE `cms_tpl_module`
+                     CHANGE `icon_list`
+                            `icon_list` VARCHAR(255) DEFAULT 'application.png' NOT NULL COMMENT 'Icon: @deprecated since 7.0.0 - no longer used - use icon_font_css_class instead'";
+TCMSLogChange::RunQuery(__LINE__, $query);
+
+$data = TCMSLogChange::createMigrationQueryData('cms_field_conf', 'en')
```
[the remark this answers is not in the document]
Comment (pixeljunkie) [reply src/CoreBundle/Bridge/Chameleon/Migration/Script/update-1554106531.inc.php]: done

File: src/CoreBundle/Bridge/Chameleon/Migration/Script/update-1554106531.inc.php [hunk 8]
```diff
@@ -0,0 +1,83 @@
...
+      'id' => '6a8eb0c7-bb06-e930-f29b-d6e9073fc1d2',
+  ])
+;
+TCMSLogChange::update(__LINE__, $data);
+
+$data = TCMSLogChange::createMigrationQueryData('cms_field_conf', 'en')
+    ->setFields([
+//        'name' => 'icon_font_css_class',
+        'translation' => 'Icon-Font CSS-Class',
+        '049_helptext' => 'This field is used for a Font-Icon. Fill in the class name here, for example for Font Awesome: fas fa-check',
```
[the remark this answers is not in the document]
Comment (pixeljunkie) [reply src/CoreBundle/Bridge/Chameleon/Migration/Script/update-1554106531.inc.php]: done

**Comment by pixeljunkie** [comment]
Added migration details to the upgrade guide and a referring info to the database update.

File: src/CoreBundle/private/modules/CMSModuleChooser/views/readonly.view.php (hunk 1 above)
Comment (pixeljunkie) [reply src/CoreBundle/private/modules/CMSModuleChooser/views/readonly.view.php]: Whoops. Forgot to select the default icon. Done.

File: src/CoreBundle/private/modules/CMSModuleChooser/views/standard.view.php (hunk 2 above)
Comment (pixeljunkie) [reply src/CoreBundle/private/modules/CMSModuleChooser/views/standard.view.php]: Fixed.

File: src/CoreBundle/Bridge/Chameleon/Migration/Script/update-1554106531.inc.php (hunk 4 above)
Comment (pixeljunkie) [reply src/CoreBundle/Bridge/Chameleon/Migration/Script/update-1554106531.inc.php]: done

File: src/CoreBundle/Resources/views/snippets/CMSModuleChooser/headerjs.html.twig (hunk 3 above)
Comment (pixeljunkie) [reply src/CoreBundle/Resources/views/snippets/CMSModuleChooser/headerjs.html.twig]: fixed

File: src/CoreBundle/Bridge/Chameleon/Migration/Script/update-1554106531.inc.php [hunk 9]
```diff
@@ -0,0 +1,83 @@
...
+TCMSLogChange::update(__LINE__, $data);
+
+$query = "ALTER TABLE `cms_tpl_module`
+                     CHANGE `icon_list`
+                            `icon_list` VARCHAR(255) DEFAULT 'application.png' NOT NULL COMMENT 'Icon: @deprecated since 7.0.0 - no longer used - use icon_font_css_class instead'";
+TCMSLogChange::RunQuery(__LINE__, $query);
+
+$recordId = TCMSLogChange::createUnusedRecordId();
+
+$data = TCMSLogChange::createMigrationQueryData('cms_field_conf', 'en')
```
Comment (umulmrum) [inline src/CoreBundle/Bridge/Chameleon/Migration/Script/update-1554106531.inc.php]:
```suggestion
$data = TCMSLogChange::createMigrationQueryData('cms_field_conf', 'de')
```

File: src/CoreBundle/Bridge/Chameleon/Migration/Service/MainMenuMigrator.php [hunk 10]
```diff
@@ -49,155 +50,8 @@ public function __construct(
...
-            'tag_blue_edit.png' => 'fas fa-tags',
-            'office-document.png' => 'fas fa-file-alt',
-            'knewsticker.png' => 'fas fa-bullhorn',
-            'kcmdf.png' => 'fas fa-object-ungroup',
-            'news.png' => 'far fa-newspaper',
-            'weather-few-clouds.png' => 'fas fa-cloud-sun',
-            'thumbnail.png' => 'fas fa-images',
-        ];
+        $iconMapping = new IconMapping();
+        $this->iconMapping = $iconMapping::ICON_MAPPING;
```
Comment (umulmrum) [inline src/CoreBundle/Bridge/Chameleon/Migration/Service/MainMenuMigrator.php]: Should be static access now.

File: src/CoreBundle/Bridge/Chameleon/Migration/Service/ModuleIconMigrator.php [hunk 11]
```diff
@@ -0,0 +1,88 @@
...
+    ) {
+        $this->databaseConnection = $databaseConnection;
+
+        $this->initIconMapping();
+    }
+
+    private function initIconMapping(): void
+    {
+        $iconMapping = new IconMapping();
+        $this->iconMapping = $iconMapping::ICON_MAPPING;
```
Comment (umulmrum) [inline src/CoreBundle/Bridge/Chameleon/Migration/Service/ModuleIconMigrator.php]: Should be static access now.

File: src/CoreBundle/Bridge/Chameleon/Migration/Script/update-1554106531.inc.php (hunk 9 above)
Comment (pixeljunkie) [reply src/CoreBundle/Bridge/Chameleon/Migration/Script/update-1554106531.inc.php]: Whoops. How did this happen? Fixed.

File: src/CoreBundle/Bridge/Chameleon/Migration/Service/MainMenuMigrator.php (hunk 10 above)
Comment (pixeljunkie) [reply src/CoreBundle/Bridge/Chameleon/Migration/Service/MainMenuMigrator.php]: Stupid me... of course... fixed.

File: src/CoreBundle/Bridge/Chameleon/Migration/Service/ModuleIconMigrator.php (hunk 11 above)
Comment (pixeljunkie) [reply src/CoreBundle/Bridge/Chameleon/Migration/Service/ModuleIconMigrator.php]: Fixed.

**Comment by umulmrum** [comment]
Thanks @pixeljunkie

**Pull request merged**